File: 1/solution.js
const fs = require('fs');
const eg = fs.readFileSync(require.resolve('./eg.txt')).toString().slice(0, -1);
const input = fs.readFileSync(require.resolve('./input.txt')).toString().slice(0, -1);

// Part 1 ---------------------------------------------------------------------

// NOTE: this is some serious spaghetti

const sortAndSeparate = input => input
    .split('\n')
    .map(i => i
        .split('   ')
        .map(i => Number(i))
    )
    .reduce(([left, right], line ) => {
        return [[...left, line[0]], [...right, line[1]]]
    }, [[], []])
    .map(list => list.sort())
    .reduce((distance, column, _, array) => {
        const [left, right] = array;

        return left.reduce((total, number, index) => {
            return total + Math.abs(number - right[index])
        }, 0)
    }, 0);

console.log('1) eg: ',    sortAndSeparate(eg));
console.log('1) input: ', sortAndSeparate(input));

// Part 2 ---------------------------------------------------------------------

const sortAndScore = input => input
    .split('\n')
    .map(i => i
        .split('   ')
        .map(i => Number(i))
    )
    .reduce(([left, right], line ) => {
        return [[...left, line[0]], [...right, line[1]]]
    }, [[], []])
    .reduce((score, column, _, array) => {
        const [left, right] = array;

        return left.reduce((probability, number) => {
            return probability + number * right.filter(i => i === number).length
        }, 0)
    }, 0);

console.log('2) eg: ',    sortAndScore(eg));
console.log('2) input: ', sortAndScore(input));

/*
Wrong guesses:

Correct:
    1) 765748
    2) 27732508
*/

File: 1/solution.py
eg = open('eg.txt').read().strip()
user_input = open('input.txt').read().strip()


# Part 1 ----------------------------------------------------------------------

def sorted_distance(user_input):
    left, right = [], []

    for line in user_input.splitlines():
        x, y = line.split()
        left.append(int(x))
        right.append(int(y))

    left.sort()
    right.sort()

    distance = 0

    for l, r in zip(left, right):
        distance += abs(l - r)

    return distance

print('1) eg: ',    sorted_distance(eg))
print('1) input: ', sorted_distance(user_input))

# Part 2 ----------------------------------------------------------------------

def similarity(user_input):
    left, right = [], []

    for line in user_input.splitlines():
        x, y = line.split()
        left.append(int(x))
        right.append(int(y))

    similarity = 0

    count = {}
    for number in left:
        if number not in count:
            count[number] = right.count(number)
        similarity += number * count[number]

    return similarity

print('2) eg: ',    similarity(eg))
print('2) input: ', similarity(user_input))

'''
Wrong guesses:

Correct:
    1) 765748
    2) 27732508
'''
